### PR TITLE
Revise Schema Loader documentation

### PR DESCRIPTION
**Pull request by josh-wong**
## Description

This PR revises the Schema Loader document.

## Related issues and/or PRs

N/A

## Changes made

- Revised contents.
- Added notice blocks for noteworthy content.
- Added content tabs for databases. 

The following is a screenshot of `schema-loader.md` on the docs site (locally).

schema-loader

## Checklist

> The following is a best-effort checklist. If any items in this checklist are not applicable to this PR or are dependent on other, unmerged PRs, please still mark the checkboxes after you have read and understood each item.

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes.
- [x] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [x] Tests (unit, integration, etc.) have been added for the changes.
- [x] My changes generate no new warnings.
- [x] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

Please see the comments in the document.

## Release notes

N/A

### CODE REVIEW

File: docs/schema-loader.md [hunk 1]
```diff
@@ -1,50 +1,83 @@
...
-  - Pass the options without a ScalarDB configuration (Deprecated).
+Since managing the schema mapping and metadata for transactions can be difficult, you can use ScalarDB Schema Loader, which is a tool to create schemas that doesn't require you to need in-depth knowledge about schema mapping or metadata.
 
-Note that this tool supports only basic options to create/delete/repair/alter a table. If you want
-to use the advanced features of a database, please alter your tables with a database specific tool after creating them with this tool.
+You have two options to specify general CLI options in Schema Loader:
 
-# Usage
+- Pass the ScalarDB properties file and database-specific or storage-specific options.
+- Pass database-specific or storage-specific options without the ScalarDB properties file. (Deprecated)
```
Comment (josh-wong) [inline docs/schema-loader.md:10]: Should we keep the deprecated database-specific options in this doc on `master` and all version branches? Or should we remove the deprecated database-specific options from all or some versions of docs?

File: docs/schema-loader.md [hunk 2]
```diff
@@ -1,50 +1,83 @@
...
+<div class="tab">
+  <button class="tablinks" onclick="openTab(event, 'Install', 'tabset-1')" id="defaultOpen-1">Install</button>
+  <button class="tablinks" onclick="openTab(event, 'Build', 'tabset-1')">Build</button>
+  <button class="tablinks" onclick="openTab(event, 'Docker', 'tabset-1')">Docker</button>
+</div>
+
+<div id="Install" class="tabcontent" markdown="1">
+
+You can download the release versions of Schema Loader from the [ScalarDB Releases](https://github.com/scalar-labs/scalardb/releases) page.
+</div>
```
Comment (josh-wong) [inline docs/schema-loader.md:34]: Since users seem to download (and later run) Schema Loader, can we change `Install` to `Download`?

File: docs/schema-loader.md [hunk 3]
```diff
@@ -1,50 +1,83 @@
...
+If you want to build Schema Loader from the source, run the following command:
 
-In case you want to build `schema-loader` from the source:
 ```console
 $ ./gradlew schema-loader:shadowJar
 ```
-- The built fat jar file is `schema-loader/build/libs/scalardb-schema-loader-<version>.jar`
 
-## Docker
+Running that command builds a fat JAR file named `scalardb-schema-loader-<VERSION>.jar` in the `schema-loader/build/libs` folder, with `<VERSION>` matching the version of ScalarDB that you are using.
```
Comment (josh-wong) [inline docs/schema-loader.md:43]: I added a short explanation about `<VERSION>` to the end of this sentence. Is this explanation correct?

File: docs/schema-loader.md [hunk 4]
```diff
@@ -182,139 +219,184 @@ Create/Delete JDBC schemas
```
Comment (josh-wong) [inline docs/schema-loader.md:219]: Should we should keep these `Deprecated` parts for these databases in the supported version branches (`3.10` to `3.5`) and on the `master` branch? Or we should completely remove these from the doc in all or some branches?

> **Note**
> 
> I haven't revised these parts yet because I wanted to confirm if we need to keep them first.

File: docs/schema-loader.md [hunk 5]
```diff
@@ -182,139 +219,184 @@ Create/Delete JDBC schemas
...
   - If `-P <CASSANDRA_PORT>` is not supplied, it defaults to `9042`.
   - If `-u <CASSANDRA_USER>` is not supplied, it defaults to `cassandra`.
   - If `-p <CASSANDRA_PASSWORD>` is not supplied, it defaults to `cassandra`.
   - `<NETWORK_STRATEGY>` should be `SimpleStrategy` or `NetworkTopologyStrategy`
 
 ```console
 # For a JDBC database
-$ java -jar scalardb-schema-loader-<version>.jar --jdbc -j <JDBC URL> -u <USER> -p <PASSWORD> -f schema.json
+$ java -jar scalardb-schema-loader-<VERSION>.jar --jdbc -j <JDBC_URL> -u <USER> -p <PASSWORD> -f schema.json
 ```
```
Comment (josh-wong) [inline docs/schema-loader.md:267]: Should we should keep these `Deprecated` parts for these databases in the supported version branches (`3.10` to `3.5`) and on the `master` branch? Or we should completely remove these from the doc in all or some branches?

> **Note**
> 
> I haven't revised these parts yet because I wanted to confirm if we need to keep them first.

File: docs/schema-loader.md [hunk 6]
```diff
@@ -182,139 +219,184 @@ Create/Delete JDBC schemas
...
 # For Cassandra
-$ java -jar scalardb-schema-loader-<version>.jar --cassandra -h <CASSANDRA_IP> [-P <CASSANDRA_PORT>] [-u <CASSANDRA_USER>] [-p <CASSANDRA_PASSWORD>] -f schema.json --alter
+$ java -jar scalardb-schema-loader-<VERSION>.jar --cassandra -h <CASSANDRA_IP> [-P <CASSANDRA_PORT>] [-u <CASSANDRA_USER>] [-p <CASSANDRA_PASSWORD>] -f schema.json --alter
 ```
 
 ```console
 # For a JDBC database
-$ java -jar scalardb-schema-loader-<version>.jar --jdbc -j <JDBC URL> -u <USER> -p <PASSWORD> -f schema.json --alter
+$ java -jar scalardb-schema-loader-<VERSION>.jar --jdbc -j <JDBC_URL> -u <USER> -p <PASSWORD> -f schema.json --alter
 ```
```
Comment (josh-wong) [inline docs/schema-loader.md:308]: Should we should keep these `Deprecated` parts for these databases in the supported version branches (`3.10` to `3.5`) and on the `master` branch? Or we should completely remove these from the doc in all or some branches?

> **Note**
> 
> I haven't revised these parts yet because I wanted to confirm if we need to keep them first.

File: docs/schema-loader.md [hunk 7]
```diff
@@ -1,50 +1,83 @@
...
-In case you want to build `schema-loader` from the source:
 ```console
 $ ./gradlew schema-loader:shadowJar
 ```
-- The built fat jar file is `schema-loader/build/libs/scalardb-schema-loader-<version>.jar`
 
-## Docker
+Running that command builds a fat JAR file named `scalardb-schema-loader-<VERSION>.jar` in the `schema-loader/build/libs` folder, with `<VERSION>` matching the version of ScalarDB that you are using.
+</div>
+<div id="Docker" class="tabcontent" markdown="1">
```
Comment (choplin) [inline docs/schema-loader.md:45]: IMO, the tabs should be in order of recommendation. From that perspective, I guess the Docker way is the most recommended since it's the easiest to start using ScalarDB. What do you think?

File: docs/schema-loader.md (hunk 4 above)
Comment (choplin) [reply docs/schema-loader.md:219]: I leave it to @brfrn169 and @feeblefakie to decide whether we should keep them. If we keep it, I think it would be good if we could hide them within the collapsible box.

File: docs/schema-loader.md [hunk 8]
```diff
@@ -182,139 +219,184 @@ Create/Delete JDBC schemas
 
 ### Create namespaces and tables
 
-For using a config file (Sample config file can be found [here](https://github.com/scalar-labs/scalardb/blob/master/conf/database.properties)):
+To create a namespace and tables by using a properties file, run the following command, replacing `<VERSION>` with the version of ScalarDB that you're using and `<PATH_TO_PROPERTIES_FILE>` with the path to the properties file. If `--coordinator` is specified, a Coordinator table will be created.
```
Comment (choplin) [inline docs/schema-loader.md:222]: I guess the `--cordinator` option should be described in the dedicated section because they affect users much. If users want to use transactions with the consensus commit (and maybe almost all ScalarDB users do), they need to create a Coordinator table. However, users would be likely to be unsure what is the Coodinator. So, IMO, it is worth emphasizing what is the Coordinator and who needs to create it here.

**Review by choplin — APPROVED** [review]
Overall, it looks good! I left several comments while approving the PR. PTAL!

File: docs/schema-loader.md (hunk 8 above)
Comment (komamitsu) [reply docs/schema-loader.md:222]: > To create a namespace and tables

Schema Loader can create [multiple namespaces](https://github.com/scalar-labs/scalardb-samples/blob/main/multi-storage-transaction-sample/schema.json), so I think "To create namespaces and tables" would be better.

File: docs/schema-loader.md [hunk 9]
```diff
@@ -182,139 +219,184 @@ Create/Delete JDBC schemas
...
+
 ```console
-$ java -jar scalardb-schema-loader-<version>.jar --config <PATH_TO_CONFIG_FILE> -f schema.json [--coordinator]
+$ java -jar scalardb-schema-loader-<VERSION>.jar --config <PATH_TO_PROPERTIES_FILE> -f schema.json [--coordinator]
```
Comment (komamitsu) [inline docs/schema-loader.md:225]: > --config <PATH_TO_PROPERTIES_FILE> -f schema.json

`schema.json` is an arbitrary file, so it should be `<PATH_TO_SCHEMA_FILE>`or something?

BTW, for schema file, the example uses abbr. name `-f` while `--config` isn't abbreviated. A bit inconsistent?

File: docs/schema-loader.md (hunk 2 above)
Comment (brfrn169) [reply docs/schema-loader.md:34]: Yes, we can change `Install` to `Download`.

File: docs/schema-loader.md [hunk 10]
```diff
@@ -1,50 +1,83 @@
...
+
+## Set up Schema Loader
+
+Select your preferred method to set up Schema Loader, and follow the instructions.
+
+<div id="tabset-1">
+<div class="tab">
+  <button class="tablinks" onclick="openTab(event, 'Install', 'tabset-1')" id="defaultOpen-1">Install</button>
+  <button class="tablinks" onclick="openTab(event, 'Build', 'tabset-1')">Build</button>
+  <button class="tablinks" onclick="openTab(event, 'Docker', 'tabset-1')">Docker</button>
```
Comment (brfrn169) [inline docs/schema-loader.md:28]: @josh-wong @feeblefakie 
I'm wondering if we really need the `Build` and `Docker` sections in this document. For the OSS documentation perspective, our target audience comprises both OSS developers and ScalarDB users, so including the `Build` and `Docker` sections might be beneficial for the developers. But for the product documentation perspective, I not sure we need the `Build` and `Docker` sections because ScalarDB users might not be interested in them.

File: docs/schema-loader.md (hunk 4 above)
Comment (brfrn169) [reply docs/schema-loader.md:219]: Since the deprecated commands will be removed in `5.0`, we need to keep the `Deprecated` parts.

**Review by brfrn169 — APPROVED** [review]
Left several comments. Please take a look when you have time!

File: docs/schema-loader.md (hunk 8 above)
Comment (josh-wong) [reply docs/schema-loader.md]: @choplin I see. How about if we direct users to [this section of the ScalarDB Java API Guide](https://scalardb.scalar-labs.com/docs/latest/api-guide/#specify-operations-for-the-coordinator-table) that briefly describes the Coordinator table?

Or, within this document, do we need to add more details about the Coordinator table that are specific to Schema Loader?

Comment (josh-wong) [reply docs/schema-loader.md]: > Schema Loader can create [multiple namespaces](https://github.com/scalar-labs/scalardb-samples/blob/main/multi-storage-transaction-sample/schema.json), so I think "To create namespaces and tables" would be better.

@komamitsu Thank you! I've applied your suggestion.

File: docs/schema-loader.md (hunk 4 above)
Comment (choplin) [reply docs/schema-loader.md:219]: > I think using tabs here and hiding the content unless the user clicks on one of them might work
I agree that this would be helpful for users, too. Thanks!

File: docs/schema-loader.md (hunk 8 above)
Comment (choplin) [reply docs/schema-loader.md]: > @choplin I see. How about if we direct users to [this section of the ScalarDB Java API Guide](https://scalardb.scalar-labs.com/docs/latest/api-guide/#specify-operations-for-the-coordinator-table) that briefly describes the Coordinator table?

Yes, this would be sufficient. We don't need to duplicate it. Thanks!

File: docs/schema-loader.md (hunk 10 above)
Comment (feeblefakie) [reply docs/schema-loader.md:28]: As @brfrn169 said, we can probably remove `Build` from the options since it's more like a user guide.
We also download/install `Docker`, so it is probably better to name them as follows?

* (Use) Fat JAR
* (Use) Docker container

Also, `Docker` should probably be more recommended in production environment, so I don't think we can say either is recommended?
What do you think? > @brfrn169

**Review by feeblefakie — COMMENTED** [review]
Overall, looking good! Thank you!
Left one comment, so PTAL!

File: docs/schema-loader.md (hunk 10 above)
Comment (brfrn169) [reply docs/schema-loader.md:28]: > Also, Docker should probably be more recommended in production environment, so I don't think we can say either is recommended?

Actually, I'm not sure which `Fat jar` or `Docker container` should be used in production environments. Considering that we also have the SQL CLI (available only in the enterprise edition) as a tool similar to the Schema Loader, it seemed to me that a 'Fat jar' might be easier for users to use. If uses can always use `Fat jar` in any environments, we can say `Fat jar` is recommended or just remove `Docker container`.

@kota2and3kan, do you have any insights on this?

File: docs/schema-loader.md (hunk 7 above)
Comment (josh-wong) [reply docs/schema-loader.md:45]: Understood. I think this will depend on which methods we decide to keep, according to the discussion in [this comment](https://github.com/scalar-labs/scalardb/pull/1196/files#r1371064322).

Considering that, I will mark this comment as resolved for now.

File: docs/schema-loader.md (hunk 2 above)
Comment (josh-wong) [reply docs/schema-loader.md]: Got it. I've changed this for now, even though keeping this part depends on how we resolve [this other comment](https://github.com/scalar-labs/scalardb/pull/1196/files#r1371064322).

File: docs/schema-loader.md (hunk 10 above)
Comment (kota2and3kan) [reply docs/schema-loader.md:28]: @feeblefakie @brfrn169  

I think we can keep both `Fat Jar` and `Docker container` and we don't need to say `which is recommended`. I think users can select out of two or more methods.

If users use `Fat Jar`, from the network perspective, they have to allow connections to access ScalarDB Cluster from some servers like the bastion server that is used for management operations. However, in the production environment, it would be better to avoid allowing unnecessary connections (i.e., other than application connections). For this reason, it might be better to use `Docker container` instead of `Fat Jar`. This is because the `Docker container` will be deployed as a pod in the Kubernetes cluster so that basically they don't need to allow additional connections other than application connections.

From this perspective (security perspective), it might be better to use the `Docker container` instead of `Fat Jar` in the production environment. But, as I mentioned, I think it is an optional thing and users can choose one of them based on their systems requirement or ScalarDB usages.

Also, as @brfrn169 mentioned, especially for commercial license users, they can use SQL CLI to create schemas and tables. In addition, users can create schemas and tables by using Admin API from their applications. These are the additional options. However, this document is a `ScalarDB Schema Loader` document. So, I think we don't need to mention SQL CLI or Admin API here.

On the other hand, I agree that the `Fat Jar` is easy to use. If we use the `Docker container` (especially in the Kubernetes environment) we need to create several manifest fils (e.g., pod, config map that includes schema configuration, and secret that includes credentials). So, ideally, I think it would be better to provide Helm Chart to use the `Docker container` like `ScalarDL Schema Loader`.

In conclusion, I think the following steps might be better.

1. In this PR, we keep `Fat Jar` and `Docker container` without any `recommendation` mark so that users can know `there are several options` and can choose options based on their requirements.
2. Create a helm chart for ScalarDB Schema Loader to use it easily on the Kubernetes environment.
3. Update the `Docker container` to `Docker container with Helm Chart` in this document.

What do you think?

Comment (kota2and3kan) [reply docs/schema-loader.md:28]: @brfrn169  
Thank you for your reply!

> I basically agree that we should offer both options, allowing users to choose based on their requirements.

Yes. I agree too.

> But what are your thoughts on the above?

Ah, sorry. I mixed some tools in my descriptions.
I agree that creating a Helm Chart for SQL CLI is not easy. We don't need to provide it.

I think we can create a Helm Chart for **ScalarDB Schema Loader**.
My image is the following.

* Custom values file

  ```yaml
  scalardbSchemaLoader:

    image:
      # For ScalarDB Schema Loader
      repository: ghcr.io/scalar-labs/scalardb-schema-loader
      # For ScalarDB Cluster Schema Loader
      repository: ghcr.io/scalar-labs/scalardb-cluster-schema-loader
  
    databaseProperties:
      # For ScalarDB Schema Loader (ScalarDB Server)
      scalar.db.transaction_manager=grpc
      scalar.db.contact_points=172.20.255.200
      scalar.db.contact_port=60051

      # For ScalarDB Cluster Schema Loader
      scalar.db.transaction_manager=cluster
      scalar.db.contact_points=direct-kubernetes:default/scalardb-cluster-envoy
  
      # Direct access to backend DB (I think this is a rare case)
      scalar.db.storage=jdbc
      scalar.db.contact_points=jdbc:postgresql://postgresql-scalardb.default.svc.cluster.local:5432/postgres
      scalar.db.username=${env:SCALAR_DB_POSTGRES_USERNAME}
      scalar.db.password=${env:SCALAR_DB_POSTGRES_PASSWORD}

    schemaJson:
      {
          "s0.t0": {
              "transaction": true,
              "partition-key": [
                  "c1"
              ],
              "clustering-key": [
                  "c2"
              ],
              "columns": {
                  "c1": "INT",
                  "c2": "INT",
                  "c3": "TEXT"
              }
          }
      }
    
    commandArgs:
      - --coordinator
      - --no-backup
      - --no-scaling
      - --repair-all 
    
    # Provide credentials via env variables like "SCALAR_DB_POSTGRES_USERNAME".
    secretName: "scalardb-credentials-secret"

    # Provide token to pull ScalarDB Cluster's Schema Loader image.
    imagePullSecrets:
      - name: secret-for-scalar
  ```

* Deploy (create schema on Kubernetes environment)

  ```console
  $ helm install scalardb-schema-loader scalar-labs/scalardb-schema-loader -f custom-values.yaml
  ```

In this way, I think users can use `ScalarDB Schema Loader` easily in a production (especially on Kubernetes) environment without additional bastion servers or allowing network connections.

So, in the future, users can see the following options on this document.

* Fat Jar
* Docker container
* Helm Chart

---

Also, I think it would be better to list the options as follows if we mention SQL CLI in this document.

* Fat Jar
* Docker container
* Helm Chart
* SQL CLI (Enterprise license only) Note: This option only includes links to SQL CLI documents.

However, I think this document is a document of ScalarDB Schema Loader. So, I think it would be better to consider whether we should mix SQL CLI in this document or not.

I think it's OK to include SQL CLI if the purpose of this document is `how to load schema`.

However, I think it's not good to include SQL CLI if the purpose of this document is `how to use ScalarDB Schema Loader`.

What do you think?

File: docs/schema-loader.md (hunk 4 above)
Comment (josh-wong) [reply docs/schema-loader.md:239]: @choplin After some testing, it looks like we can't hide the contents of a tab. In the following screenshot, when the user selects `Cassandra`, the content doesn't appear as expected (other tabs work fine though). Also, even if we could, the tabbed section looks a little awkward without any content showing:

tabbed-content-initial-view-disabled

For now, I've implemented the following style that uses tabs in a **Note** block:

tabbed-content-initial-view-enabled-in-note-block

Do you have any concerns about this method of showing deprecated content?

File: docs/schema-loader.md (hunk 10 above)
Comment (brfrn169) [reply docs/schema-loader.md:28]: As discussed offline with @kota2and3kan, we concluded that we keep both `Fat Jar` and `Docker container`, and in the future we can add `Helm Chart` (maybe as enterprise only) in addition to them.

@josh-wong Can you please delete `Build` and keep `Fat Jar` and `Docker container`? And we don't need to say `which is recommended`.

File: docs/schema-loader.md (hunk 9 above)
Comment (komamitsu) [reply docs/schema-loader.md]: Oh, I didn't know we often used `--config` and `-f`. Your idea sounds good to me. Let's leave it as is in this PR.

**Review by komamitsu — APPROVED** [review]
LGTM! Thank you!

File: docs/schema-loader.md (hunk 10 above)
Comment (josh-wong) [reply docs/schema-loader.md]: @brfrn169 @kota2and3kan Thank you for confirming which methods to keep! I've removed the `Build` content.

In the Docker section, we also mention the following. Can we keep this part or should we remove it?:

````markdown
You can also build the Docker image as follows by running the following command:

```console
$ ./gradlew schema-loader:docker
```
````

Comment (brfrn169) [reply docs/schema-loader.md]: @josh-wong Thank you for removing the `Build` content.

> In the Docker section, we also mention the following. Can we keep this part or should we remove it?:

I think we should remove this because it's likely not of interest to users. Thanks.

**Review by brfrn169 — APPROVED** [review]
Other than what I commented, LGTM! Thank you!

File: docs/schema-loader.md (hunk 10 above)
Comment (josh-wong) [reply docs/schema-loader.md]: Thank you for confirming! I've removed that part.

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!
